Bugfix for ConfiguredObjectProvenance and StructuralEquals on ConfigurationData

JackSullivan opened the pull request:
## Bugfix

Provenance instances for implementations that depend on `SkeletalConfiguredObjectProvenance` populate an empty string rather than nothing for null-valued string fields that are annotated with `@Config`. This fixes that issue. The issue was discovered because it was causing `ConfigurationData` instances produced through the provenance system to not match those produced through importing a configurable object into `ConfigurationManager`. There may be instances of incorrect provenances in the wild that this affects, but I suspect the chances are fairly small.

## Enhancement

Added a method to check whether two configured objects are structurally equal to one another by checking whether their types and property values match, dereferencing property values that correspond to the names of other configured objects and recursively equality checking them.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
+     *
+     * <p></p>
```
Can you remove the `</p>` close tag?

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+     * in {@code b}. {@code a} and {@code b} should each contain all the ConfigurationData objects
+     * needed to instantiate the objects named by {@code aName} and {@code bName} respectively. Objects not
+     * instantiated by traversing children of {@code aName} and {@code bName} are ignored.
+     *
```
Might be useful to note that this method is extremely chatty at the info level in the Javadoc when it finds discrepancies. Most other things in OLCUT tend to log at fine unless it's an actual issue, and that might be appropriate here.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+                                        aMap.keySet().stream().allMatch(k -> {
+                                            boolean eq = simplePropertyDerefEquals(a, b, aMap.get(k), bMap.get(k), propName);
+                                            if(!eq) {
+                                                logger.info(String.format("Property key: %s, PropValue Key: %s, a.value: %s, b.value: %s",
```
Doesn't `simplePropertyDerefEquals` already log this? If so, then this lambda can drop the return statement and it'll be much nicer.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+
+            return typesMatch &&
+                    propNames.stream().allMatch(propName -> {
+                        Optional<Property> aPropOpt = Optional.ofNullable(aRoot.getProperties().get(propName));
```
Can we extract this lambda into a static method? It's hard to figure out what it's capturing and that level of complexity in a lambda body worries me (e.g. it's harder to debug). If possible could you recursively do that so the nested lambdas are also static methods? There are too many return statements for too many functions.

Craigacp reviewed (changes requested):
Could you also add a test where the equals is true and one where it's false? You should be able to generate objects which round trip through provenance using the existing provenance test utils and then compare to see if they are equal.

JackSullivan commented:
I believe I've addressed all of the changes addressed above. In the process I removed the tests that used `Random` fields throughout, since they are both deprecated and incompatible with provenance.

JackSullivan commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +274,152 @@ public int hashCode() {
+                Map<String, SimpleProperty> bMap = ((MapProperty) bProp).getMap();
+                return aMap.keySet().equals(bMap.keySet()) &&
+                        aMap.keySet().stream().allMatch(k -> simplePropertyDerefEquals(a, b, aMap.get(k), bMap.get(k), propName, Optional.of(", PropValue key: " + k)));
+            } else {
```
Add a clause to account for mismatched property types

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                                return eq;
+                            });
+
+                if(!listMatch) {
```
This line doesn't seem to be indented properly.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                    boolean listMatch = checkPresenceAllMatch(this.listProperties,that.listProperties, (as, bs) -> {
+                       boolean eq = Util.bagEquality(as, bs);
+                        if(!eq) {
+                            logger.fine("bags not equal:\na: " + as.toString() +  "\nb: " + bs.toString());
```
I think this should be `ListProperties not equal`, or `ListProperties not equal using bag equality`. I see there's more logging further down, but I think being specific here and in the class one will make it a little easier to see.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+                    innerBool.map(d -> "(parsed as bool " + d + ")").orElse("(not parsed as bool)"));
+        }
+
+        @Override public int hashCode() {
```
We have override on the line before the method declaration.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+     * Checks whether two ConfigurationData objects are 'structurally equal'. Two objects are structurally
+     * equal when their classNames are the same and when all of their properties are equal or, if those
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
```
Add a note saying it does not compare the serializedForm, importable, exportable, leaseTime or entries attributes of the configuration data. Everything apart from serializedForm is going to go away in the next version anyway, but it's worth noting here.

Craigacp commented on olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
+     *
+     * <p>
+     *
+     * At log-level {@code FINE} this reports the first place where the two instances differ, and the nature of their
```
Maybe expand this slightly? It logs all the places where the two instances differ in whatever object they differ in. Possibly `reports the first object where the two instances differ`, but that's not particularly clear either.

Craigacp commented on olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java:
```diff
@@ -0,0 +1,96 @@
+package com.oracle.labs.mlrg.olcut.config;
```
This file needs a copyright & license header.

Craigacp commented on olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java:
```diff
@@ -0,0 +1,96 @@
+    public void structuralEqualsAllConfig() {
+
+        ConsoleHandler ch = new ConsoleHandler();
+        ch.setLevel(Level.FINE);
```
Can you back out the logging changes from this test?

Craigacp reviewed (changes requested):
A bunch of small things, mostly formatting and docs.

Craigacp reviewed (approved):
LGTM
